### PR TITLE
perf(sm120): use native FlashInfer FA2 for prefill

### PR DESCRIPTION
**Pull request by HJCheng0602**
Reuse the existing fused RoPE and KV-cache path while selecting the faster native ragged FA2 kernel for supported prefix-free prefill workloads.

### CODE REVIEW

File: rtp_llm/models_py/modules/factory/attention/cuda_impl/trt.py [hunk 1]
```diff
@@ -486,3 +489,79 @@ def prepare_cuda_graph(self, attn_inputs: PyAttentionInputs) -> None:
...
+
+    def __init__(
+        self,
+        attn_configs: AttentionConfigs,
+        attn_inputs: PyAttentionInputs,
+        parallelism_config: Optional[ParallelismConfig] = None,
+    ) -> None:
+        self.need_rope = attn_configs.rope_config.style != RopeStyle.No
+        self.attn_configs = attn_configs
+        self.fmha_impl = PyFlashinferPrefillAttnOp(attn_configs, backend="fa2")
```
Comment (LLLLKKKK) [inline rtp_llm/models_py/modules/factory/attention/cuda_impl/trt.py]: **[P1] Native FA2 路径丢失自定义 softmax scale**

原 `TRTLLMFMHAv2PrefillOp` 使用 `softmax_extra_scale / q_scaling * head_dim^-0.5`，而新候选创建 `PyFlashinferPrefillAttnOp` 后未向 `plan()` 传递 `sm_scale`，因此使用默认 `head_dim^-0.5`。任一缩放配置非 1 时，新优先路径都会静默改变 attention 概率和生成结果。

**建议:** 显式传递等价的 `sm_scale`；若暂不支持，应在 `support()` 中拒绝非默认缩放并回退，同时增加非默认 `q_scaling` 和 `softmax_extra_scale` 的数值测试。

**Checklist:** [6.1] LSP：子类/重写保持基类契约；[6.1] 新逻辑有聚焦单测 + 相关集成/smoke 测试

**Review by LLLLKKKK — DISMISSED** [review]
## AI Code Review - PR #1380

**Status:** BLOCKING

**Summary:** P0/0 · P1/1 · P2/1 · P3/1

**Reviewed:** commit `b7b51422f7eb` · 2026-09-04 11:44 UTC+8

### Blocking Issues

#### P1
- **Native FA2 路径丢失自定义 softmax scale** @ `rtp_llm/models_py/modules/factory/attention/cuda_impl/trt.py:510`
  - 建议：显式传递等价的 `sm_scale`；若暂不支持，应在 `support()` 中拒绝非默认缩放并回退，同时增加非默认 `q_scaling` 和 `softmax_extra_scale` 的数值测试。

### Non-blocking Suggestions

#### P2
- **SM90 FP8 被错误声明为 FA2 支持** @ `rtp_llm/models_py/modules/factory/attention/cuda_impl/trt.py:524`
  - 建议：若优化仅面向 SM120，应增加 `is_sm12x()` 限制；否则按架构和 dtype 选择 FA2/FA3，并测试 SM90 BF16、SM90 FP8 与 SM120 的工厂选择和输出。

#### P3
- **CUDA Graph 回退前会常驻无用的 128 MiB workspace** @ `rtp_llm/models_py/modules/factory/attention/cuda_impl/trt.py:524`
  - 建议：在静态 `support()` 中直接拒绝 CUDA Graph 输入，并增加断言确认图模式不会创建该 wrapper 或扩大 FlashInfer workspace 池。

### Checklist Findings (5 fail / 95 total)

**General Principles Checklist**
- [6.1] Architecture — 状态不变量：创建/更新/失败/重试/回滚路径有效  → issue `CUDA Graph 回退前会常驻无用的 128 MiB workspace`
  _`support()` 未检查 `attn_inputs.is_cuda_graph`，工厂会先构造实例并执行 FlashInfer plan；此过程申请 128 MiB workspace。随后 `support_cuda_graph()` 返回 False，实例被放弃，但析构会将缓冲区放入全局池，导致图模式常驻一块从未用于推理的显存。_
- [6.1] Architecture — 错误语义：fail-fast/retry/fallback/silent 行为显式  → issue `SM90 FP8 被错误声明为 FA2 支持`
  _`_supports_trtllm_fmha_v2()` 接受 SM90 FP8 KV cache，新候选却固定 `backend="fa2"`；同时 `attn_q_dtype()` 会在该组合下生成仅适用于 FA3 的 FP8 Q。构造阶段的 FA2 plan 因而失败，工厂捕获异常后才回退，造成错误的支持声明、逐层告警和无效初始化；SM90 BF16 也会被无意切换默认后端。_
- [6.1] Software Engineering — LSP：子类/重写保持基类契约  → issue `Native FA2 路径丢失自定义 softmax scale`
  _原 `TRTLLMFMHAv2PrefillOp` 使用 `softmax_extra_scale / q_scaling * head_dim^-0.5`，而新候选创建 `PyFlashinferPrefillAttnOp` 后未向 `plan()` 传递 `sm_scale`，因此使用默认 `head_dim^-0.5`。任一缩放配置非 1 时，新优先路径都会静默改变 attention 概率和生成结果。_
- [6.1] Tests — 分布式/跨平台变更有对应覆盖  → issue `SM90 FP8 被错误声明为 FA2 支持`
  _`_supports_trtllm_fmha_v2()` 接受 SM90 FP8 KV cache，新候选却固定 `backend="fa2"`；同时 `attn_q_dtype()` 会在该组合下生成仅适用于 FA3 的 FP8 Q。构造阶段的 FA2 plan 因而失败，工厂捕获异常后才回退，造成错误的支持声明、逐层告警和无效初始化；SM90 BF16 也会被无意切换默认后端。_
- [6.1] Tests — 新逻辑有聚焦单测 + 相关集成/smoke 测试  → issue `Native FA2 路径丢失自定义 softmax scale`
  _原 `TRTLLMFMHAv2PrefillOp` 使用 `softmax_extra_scale / q_scaling * head_dim^-0.5`，而新候选创建 `PyFlashinferPrefillAttnOp` 后未向 `plan()` 传递 `sm_scale`，因此使用默认 `head_dim^-0.5`。任一缩放配置非 1 时，新优先路径都会静默改变 attention 概率和生成结果。_

### Strengths
- 复用了现有 fused RoPE、KV-cache 写入和 FlashInfer ragged attention 组件。
- 沿用 `disable_flashinfer_native` 回滚开关，并保留后备实现。

File: rtp_llm/models_py/modules/factory/attention/cuda_impl/trt.py [hunk 2]
```diff
@@ -486,3 +489,79 @@ def prepare_cuda_graph(self, attn_inputs: PyAttentionInputs) -> None:
...
+        self.write_cache_store_impl = common.create_write_cache_store_impl(attn_inputs)
+
+    @classmethod
+    def support(
+        cls, attn_configs: AttentionConfigs, attn_inputs: PyAttentionInputs
+    ) -> bool:
+        # Mirrors the TRT FMHA v2 arch/dtype/head-dim gate so this only replaces
+        # the attention kernel on configurations already known to work, and adds
+        # the native op's own ragged (prefix-free) requirement.
+        return _supports_trtllm_fmha_v2(
```
Comment (LLLLKKKK) [inline rtp_llm/models_py/modules/factory/attention/cuda_impl/trt.py]: **[P2] SM90 FP8 被错误声明为 FA2 支持**

`_supports_trtllm_fmha_v2()` 接受 SM90 FP8 KV cache，新候选却固定 `backend="fa2"`；同时 `attn_q_dtype()` 会在该组合下生成仅适用于 FA3 的 FP8 Q。构造阶段的 FA2 plan 因而失败，工厂捕获异常后才回退，造成错误的支持声明、逐层告警和无效初始化；SM90 BF16 也会被无意切换默认后端。

**建议:** 若优化仅面向 SM120，应增加 `is_sm12x()` 限制；否则按架构和 dtype 选择 FA2/FA3，并测试 SM90 BF16、SM90 FP8 与 SM120 的工厂选择和输出。

**Checklist:** [6.1] 错误语义：fail-fast/retry/fallback/silent 行为显式；[6.1] 分布式/跨平台变更有对应覆盖

Comment (LLLLKKKK) [inline rtp_llm/models_py/modules/factory/attention/cuda_impl/trt.py]: **[P3] CUDA Graph 回退前会常驻无用的 128 MiB workspace**

`support()` 未检查 `attn_inputs.is_cuda_graph`，工厂会先构造实例并执行 FlashInfer plan；此过程申请 128 MiB workspace。随后 `support_cuda_graph()` 返回 False，实例被放弃，但析构会将缓冲区放入全局池，导致图模式常驻一块从未用于推理的显存。

**建议:** 在静态 `support()` 中直接拒绝 CUDA Graph 输入，并增加断言确认图模式不会创建该 wrapper 或扩大 FlashInfer workspace 池。

**Checklist:** [6.1] 状态不变量：创建/更新/失败/重试/回滚路径有效

**Review by LLLLKKKK — COMMENTED** [review]
## AI Code Review - PR #1380 (non-blocking suggestions)

2 条 P2/P3 建议，不阻塞合并。阻塞判定与完整摘要见上一条 review。

File: rtp_llm/models_py/modules/factory/attention/cuda_impl/test/test_trtllm_fmha_v2_prefill.py [hunk 3]
```diff
@@ -17,12 +21,99 @@
...
+        return config
+
+    @staticmethod
+    def _inputs(prefix_lengths=(0,), is_cuda_graph=False) -> PyAttentionInputs:
+        inputs = PyAttentionInputs()
+        inputs.prefix_lengths = torch.tensor(prefix_lengths, dtype=torch.int32)
+        inputs.is_cuda_graph = is_cuda_graph
+        return inputs
+
+    def test_supports_validated_sm12x_shape(self):
```
Comment (LLLLKKKK) [inline rtp_llm/models_py/modules/factory/attention/cuda_impl/test/test_trtllm_fmha_v2_prefill.py:57]: **[P2] 新增测试绕过工厂，无法验证实际选型与回退**

支持矩阵直接调用 `support()`，开关测试仅调用 `_is_fmha_impl_disabled()`，数值测试也直接构造实现；没有测试经过 `get_fmha_impl()`。因此漏注册、优先级错误或关闭后的回退链异常时，新增测试仍可能全部通过。

**建议:** 增加工厂级测试，验证合法 SM12x 配置选择 Native FA2，并验证关闭开关、CUDA Graph 和不支持配置会回退至预期实现。

**Checklist:** [6.1] 新逻辑有聚焦单测 + 相关集成/smoke 测试

File: rtp_llm/models_py/modules/factory/attention/__init__.py [hunk 4]
```diff
@@ -83,6 +83,7 @@
             HeadWisePrefillImpl,
             FlashInferTRTLLMSpecDecodeImpl,
             FlashInferTRTLLMPrefillImpl,
+            FlashInferNativePrefillImpl,
```
Comment (LLLLKKKK) [inline rtp_llm/models_py/modules/factory/attention/__init__.py:86]: **[P2] 新 FA2 prefill 路径缺少可安全使用的独立回滚开关**

新候选位于 TRT FMHA v2 之前，但唯一门控 `disable_flashinfer_native` 也会禁用 SM12x decode 所需的 `PyFlashinferDecodeImpl`；TRT-LLM decode 和 XQA 均拒绝 SM12x。因此回滚新 prefill 路径会使 decode 无可用实现。

**建议:** 增加仅控制 `FlashInferNativePrefillImpl` 的配置项，并验证关闭后 prefill 回退至 TRT FMHA v2、decode 仍选择 `PyFlashinferDecodeImpl`。

**Checklist:** [6.1] 回滚路径：风险行为存在运维回滚手段

File: rtp_llm/models_py/modules/factory/attention/cuda_impl/trt.py [hunk 5]
```diff
@@ -486,3 +489,94 @@ def prepare_cuda_graph(self, attn_inputs: PyAttentionInputs) -> None:
...
+        self.attn_inputs = attn_inputs
+        self.fmha_impl.prepare(attn_inputs)
+        self.rope_params = self.rope_kvcache_impl.prepare(attn_inputs)
+        self.write_cache_store_impl = common.create_write_cache_store_impl(attn_inputs)
+
+    @classmethod
+    def support(
+        cls, attn_configs: AttentionConfigs, attn_inputs: PyAttentionInputs
+    ) -> bool:
+        if getattr(attn_inputs, "is_cuda_graph", False):
```
Comment (LLLLKKKK) [inline rtp_llm/models_py/modules/factory/attention/cuda_impl/trt.py:528]: **[P3] CUDA Graph 门控绕过已声明的静态字段**

`PyAttentionInputs.is_cuda_graph` 已在类型桩声明且由工厂在调用 `support()` 前赋值；使用 `getattr(..., False)` 会绕过静态检查，并在字段契约异常时静默按非图模式选择 Native FA2。

**建议:** 直接访问 `attn_inputs.is_cuda_graph`，使字段契约缺失时显式失败。

**Checklist:** [P.A] 禁止 getattr/setattr literal 访问

**Review by LLLLKKKK — COMMENTED** [review]
## AI Code Review - PR #1380

**Status:** LGTM

**Summary:** P0/0 · P1/0 · P2/2 · P3/1

**Reviewed:** commit `d826e583d988` · 2026-09-04 13:22 UTC+8

lgtm ready to ci

### Non-blocking Suggestions

#### P2
- **新增测试绕过工厂，无法验证实际选型与回退** @ `rtp_llm/models_py/modules/factory/attention/cuda_impl/test/test_trtllm_fmha_v2_prefill.py:57`
  - 建议：增加工厂级测试，验证合法 SM12x 配置选择 Native FA2，并验证关闭开关、CUDA Graph 和不支持配置会回退至预期实现。
- **新 FA2 prefill 路径缺少可安全使用的独立回滚开关** @ `rtp_llm/models_py/modules/factory/attention/__init__.py:86`
  - 建议：增加仅控制 `FlashInferNativePrefillImpl` 的配置项，并验证关闭后 prefill 回退至 TRT FMHA v2、decode 仍选择 `PyFlashinferDecodeImpl`。

#### P3
- **CUDA Graph 门控绕过已声明的静态字段** @ `rtp_llm/models_py/modules/factory/attention/cuda_impl/trt.py:528`
  - 建议：直接访问 `attn_inputs.is_cuda_graph`，使字段契约缺失时显式失败。

### Checklist Findings (3 fail / 88 total)

**General Principles Checklist**
- [6.1] Architecture — 回滚路径：风险行为存在运维回滚手段  → issue `新 FA2 prefill 路径缺少可安全使用的独立回滚开关`
  _新候选位于 TRT FMHA v2 之前，但唯一门控 `disable_flashinfer_native` 也会禁用 SM12x decode 所需的 `PyFlashinferDecodeImpl`；TRT-LLM decode 和 XQA 均拒绝 SM12x。因此回滚新 prefill 路径会使 decode 无可用实现。_
- [6.1] Tests — 新逻辑有聚焦单测 + 相关集成/smoke 测试  → issue `新增测试绕过工厂，无法验证实际选型与回退`
  _支持矩阵直接调用 `support()`，开关测试仅调用 `_is_fmha_impl_disabled()`，数值测试也直接构造实现；没有测试经过 `get_fmha_impl()`。因此漏注册、优先级错误或关闭后的回退链异常时，新增测试仍可能全部通过。_

**Python Static-First Checklist**
- [P.A] 静态结构与类型纪律 — 禁止 getattr/setattr literal 访问  → issue `CUDA Graph 门控绕过已声明的静态字段`
  _`PyAttentionInputs.is_cuda_graph` 已在类型桩声明且由工厂在调用 `support()` 前赋值；使用 `getattr(..., False)` 会绕过静态检查，并在字段契约异常时静默按非图模式选择 Native FA2。_

### Strengths
- Native FA2 严格限定于 SM12x、BF16、BASE KV cache、因果 GQA、head_dim 256 和非 CUDA Graph 场景。
- 自定义 softmax scale 已完整传入 FlashInfer plan，并覆盖两种 RoPE 配置的数值一致性测试。
- CUDA Graph 在实现构造前即被拒绝，避免分配无用 workspace。
